### PR TITLE
Implemented support for extensions

### PR DESCRIPTION
**Pull request by GrahamCampbell**

### CODE REVIEW

File: src/CloudEvents/V1/CloudEvent.php [hunk 1]
```diff
@@ -145,4 +164,47 @@ public function setTime(?DateTimeInterface $time): CloudEvent
...
+    }
+
+    /**
+     * @param bool|int|string|null $value
+     */
+    public function setExtension(string $attribute, $value): CloudEvent
+    {
+        if (\preg_match('/^[a-z0-9]+$/', $attribute) !== 1) {
+            throw new ValueError(
+                \sprintf('%s(): Argument #1 ($attribute) must match the regex [a-z0-9]+, %s given', __METHOD__, $attribute)
```
Comment (GrahamCampbell) [inline src/CloudEvents/V1/CloudEvent.php:183]: NB The spec says MUST match `[a-z0-9]+` and SHOULD be no longer than 20 characters.

File: src/CloudEvents/V1/CloudEvent.php [hunk 2]
```diff
@@ -145,4 +164,47 @@ public function setTime(?DateTimeInterface $time): CloudEvent
...
+                \sprintf('%s(): Argument #1 ($attribute) must not be a reserved attribute, %s given', __METHOD__, $attribute)
+            );
+        }
+
+        $type = \get_debug_type($value);
+        $types = ['bool', 'int', 'string', 'null'];
+
+        if (!in_array($type, $types, true)) {
+            throw new TypeError(
+                \sprintf('%s(): Argument #2 ($value) must be of type %s, %s given', __METHOD__, implode('|', $types), $type)
```
Comment (GrahamCampbell) [inline src/CloudEvents/V1/CloudEvent.php:198]: Notice `DateTimeImmutable` is not supported. This is in line with the Java SDK. The reason for that behaviour is that there's no way to tell the difference between a string that looks like a datetime and an actual datetime. We thus leave it to the user to decode and encode datetimes (possibly using the stuff in `CloudEvents\Serializers\Formatters`) before providing using them as extensions.

File: tests/CloudEvents/V1/CloudEventTest.php [hunk 3]
```diff
@@ -114,6 +116,70 @@ public function testGetSetData(): void
...
+        $this->expectException(TypeError::class);
+
+        $event->setExtension('comacme', 1.1);
+    }
+
+    /**
+     * @covers ::setExtension
+     * @covers ::getExtensions
+     */
+    public function testCabSetAndUnsetExtensions(): void
```
Comment (jlaswell) [inline tests/CloudEvents/V1/CloudEventTest.php]: 🚕

```suggestion
    public function testCanSetAndUnsetExtensions(): void
```

**Review by jlaswell — APPROVED** [review]
Single typo, but the implementation here looks good.

**Pull request merged**